awx/api/serializers.py

```diff
@@ -1959,6 +1964,137 @@ def to_representation(self, obj):
         return ret
 
 
+class BulkHostSerializer(HostSerializer):
+    name = serializers.CharField(required=True, allow_blank=False, max_length=512)
+    instance_id = serializers.CharField(required=False, max_length=1024)
+    description = serializers.CharField(required=False)
+    enabled = serializers.BooleanField(default=True, required=False)
+    variables = serializers.CharField(write_only=True, required=False)
+
+    class Meta:
+        fields = (
+            'name',
+            'enabled',
+            'instance_id',
+            'description',
+            'variables',
+        )
+
+
+class BulkHostCreateSerializer(serializers.Serializer):
+    inventory = serializers.PrimaryKeyRelatedField(
+        queryset=Inventory.objects.all(), required=True, write_only=True, help_text=_('Primary Key ID of inventory to add hosts to.')
+    )
+    hosts_help_text = _('List of hosts to be created, JSON. e.g. [{"name": "example.com"}, {"name": "127.0.0.1"}]')
+    hosts = serializers.ListField(
+        child=BulkHostSerializer(), allow_empty=False, max_length=settings.BULK_HOST_MAX_CREATE, write_only=True, help_text=hosts_help_text
+    )
+
+    class Meta:
+        model = Inventory
+        fields = ('inventory', 'hosts')
+        read_only_fields = ()
+
+    def raise_if_cannot_add_hosts(self, attrs):
+        validation_info = get_licenser().validate()
+
+        org = attrs['inventory'].organization
+
+        if org:
+            org_active_count = Host.objects.org_active_count(org.id)
+            new_hosts = [h['name'] for h in attrs['hosts']]
+            org_net_new_host_count = Host.objects.filter(inventory__organization=org.id).exclude(name__in=new_hosts).count()
```

Comment:
This name `org_net_new_host_count` doesn't jive with the query in my head. The new hosts, excluding name conflicts, would be `len(new_hosts)`, which is a list of host names. This query is counting hosts in the database, excluding any names in the new set of hosts. The name suggests that it is a count of the new hosts, excluding those already in the database (in the organization).

So what I read from the name would suggest

```python
org_net_new_host_count = len(new_hosts) - Host.objects.filter(inventory__organization=org.id).exclude(name__in=new_hosts).count()
```

But even this falls short, because of the corner case where a new host has a name which is present in _two_ existing hosts in the organization. This is why we do the `order_by` and `distinct` in the host count stuff. You could add this to make the above plausibly correct.

Reply:
@AlanCoding Great catch ! I edited it to be like this now 

```
org_net_new_host_count = len(new_hosts) - Host.objects.filter(inventory__organization=1, name__in=new_hosts).values('name').distinct().count()
```

Example:

```
In [21]: Host.objects.filter(inventory__organization=1).values('name')
Out[21]: <QuerySet [{'name': 'example1.com'}, {'name': 'example2.com'}, {'name': 'host1'}, {'name': 'host1'}, {'name': 'host2'}, {'name': 'host3'}, {'name': 'host4'}, {'name': 'host5'}, {'name': 'host6'}, {'name': 'localhost'}]>

In [22]: new_hosts = ["host1"]

In [23]: Host.objects.filter(inventory__organization=1, name__in=new_hosts).values('name').distinct().count()
Out[23]: 1

In [24]: len(new_hosts) - Host.objects.filter(inventory__organization=1, name__in=new_hosts).values('name').distinct().count()
Out[24]: 0
```

Reply:
good catch yall :D